backends/ebpf/psa/README.md

```diff
@@ -0,0 +1,264 @@
+
+By default, a packet is further passed to the TC subsystem. It is done by `XDP_PASS` action and packet is further handled by `tc-ingress` program.
+
+## RESUBMIT
+
+The purpose of `RESUBMIT` is to transfer packet processing back to the Ingress Parser from Ingress Deparser.
+
+We implement packet resubmission by calling main `ingress()` function (implementing the PSA Ingress pipeline) in a loop. 
+The `MAX_RESUBMIT_DEPTH` variable specifies maximum number of resubmit operations (the `MAX_RESUBMIT_DEPTH` value is currently hardcoded and is set to 4).
+The `resubmit` flag defines whether the `tc-ingress` program should enter next iteration (resubmit)
+or break the loop. The pseudocode looks as follows:
+
+```c
+int i = 0;
+int ret = TC_ACT_UNSPEC;
+for (i = 0; i < MAX_RESUBMIT_DEPTH; i++) {
+    out_md.resubmit = 0;
+    ret = ingress(skb, &out_md);
+    if (out_md.resubmit == 0) {
+        break;
+    }
+}
+```
+
+## NU (Normal Unicast), NM (Normal Multicast), CI2E (Clone Ingress to Egress)
+
+NU, NM and CI2E refer to process of sending packet from the PSA Ingress Pipeline (more specifically from the Traffic Manager)
+to the PSA Egress pipeline. The NU path is implemented in the eBPF subsystem by invoking the `bpf_redirect()` helper from
+the `tc-ingress` program. This helper sets an output port for a packet and the packet is further intercepted by the TC egress.
+
+Both NM and CI2E require the `bpf_clone_redirect()` helper to be used. It redirects a packet to an output port, but also
+clones a packet buffer, so that a packet can be copied and sent to multiple interfaces. From the eBPF program's perspective, 
+`bpf_clone_redirect()` must be invoked in the loop to send packets to all ports from a clone session/multicast group.
+
+Clone sessions or multicast groups and theirs members are stored as a BPF array map of maps (`BPF_MAP_TYPE_ARRAY_OF_MAPS`). 
+The P4-eBPF compiler generates two outer BPF maps: `multicast_grp_tbl` and `clone_session_tbl`. Both of them store inner maps indexed by 
+the clone session or multicast group identifier, respectively. The clone session/multicast group members (defining `egress_port`, `instance` or other parameters used by clone sessions)
+are stored in the inner hash map. 
+
+While performing the packet replication, the eBPF program does a lookup to the outer map based on the clone session/multicast group identifier and, then,
+does another lookup to the inner map to find all members.
+
+## CE2E (Clone Egress to Egress)
+
+CE2E refers to process of copying a packet that was handled by the Egress pipeline and resubmitting the cloned packet to the Egress Parser.
+
+CE2E is implemented by invoking `bpf_clone_redirect()` helper in the Egress path. Output ports are determined based on the
+`clone_session_id` and lookup to "clone_session" BPF map, which is shared among TC ingress and egress (eBPF subsystem allows for map sharing between programs).
+
+## Sending packet to CPU
```

Comment:
The subtitle "Sending packet to CPU" is slightly confusing in the context of eBPF because everything is running on CPU. Perhaps it would be more appropriate to change the subtitle to "Sending packet to Control Plane"?

@jafingerhut What do you think?

Reply:
The current PSA specification might be contributing to some confusion here, since that spec does refer to a "CPU port" to send packets from the data plane to the controller.  That specification does have at least several places with a kind of built-in assumption that the P4 data plane is separate from a nearby control CPU, at least in how it names things.

I'm not sure if there are only a few small surgical changes in the PSA spec that would match your suggestion here, but unfortunately if we want to avoid changing names in the P4 include file for the architecture, there are already a few names like these in there (https://github.com/p4lang/p4c/blob/main/p4include/bmv2/psa.p4):

```
const PortId_t PSA_PORT_CPU = (PortId_t) 0xfffffffd;

const CloneSessionId_t PSA_CLONE_SESSION_TO_CPU = (CloneSessionId_t) 0;
```